File: src_julia/epanet/epanet.jl
# Módulo para rodar simulação
module epanet

include("../epamodule/epamodule.jl")
em = Main.epamodule

using DataFrames
using CSV

# exporting structs
export SimulationValues, Paths

# exporting functions



# struct para controle de paths
struct Paths
    nodes::String
    links::String
    inp::String
    values::String
end # end struct Paths


struct SimulationValues
    real_values::DataFrame
    target::Array{Float64, 1}
    all_nodes::Array{Int64,1}
    g_links::Array{Array{Int64,1},1}
    dim::Int64
    paths::Paths
    function SimulationValues(paths::Paths,
        target::Array{Float64, 1}, 
        nos_dim::Int64, 
        posicao::Float64 = 0.1, 
        vazoes::Array{Int64, 1} = [20,30,50,55,60,70] )
        # Puxando valores
        dim = length(target)
        tubulacoes = split.(read(paths.links, String), "\n")
        start_sim(paths)
        all_nodes = get_allnodes(paths)        
        t_nodes = get_target_nodes(nos_dim, tubulacoes, posicao)
        g_links = get_groups(dim, tubulacoes)
        get_real_values(paths, vazoes, target, t_nodes,g_links, all_nodes)
        real_values = CSV.File(paths.values) |> DataFrame
        close_sim()
        new(real_values, target, all_nodes, g_links, dim, paths)
    end 
end


function release_net(n)::Nothing
    global net = n
end

# Função que começa simulação do epanet
function start_sim(paths::Paths)::Nothing
    em.ENopen(paths.inp)
    em.ENopenH()
end

# Função que fecha simulação do epanet
function close_sim()::Nothing
    em.ENcloseH()
    em.ENclose()
end

function restart_sim(paths::Paths)::Nothing
    close_sim()
    start_sim(paths)
end


# Função que lê arquivo dos nós e obtém os IDs de todos eles
function get_allnodes(paths::Paths)::Array{Int64}
    arq = open(paths.nodes)
    nodes = em.ENgetnodeindex.(string.(split(read(arq,String),"\n")))
    return nodes
end


function update_network_values(values, net)::Nothing
    for i in 1:length(values)
        muda_rugosidade.(net.g_links[i],values[i])
    end
    em.ENsolveH()
end

function update_network_values(values::Array{Float64,1}, l_groups::Array{Array{Int64,1},1})::Nothing
    for i in 1:length(values)
        muda_rugosidade.(l_groups[i],values[i])
    end
    em.ENsolveH()
end


# function muda_vazao(valor, net::SimulationValues)::Nothing
#     em.ENsetnodevalue(net.all_nodes, em.EN_BASEDEMAND, em.ENgetnodevalue(node, em.EN_BASEDEMAND)*valor)
# end

function muda_vazao(valor, node)::Nothing
    em.ENsetnodevalue(node, em.EN_BASEDEMAND, em.ENgetnodevalue(node, em.EN_BASEDEMAND)*valor |> Float64)
end

# function reverte_vazao(valor, net::SimulationValues)::Nothing
#     em.ENsetnodevalue(net.all_nodes, em.EN_BASEDEMAND, em.ENgetnodevalue(node, em.EN_BASEDEMAND)/valor)
# end

function reverte_vazao(valor, nodes)::Nothing
    em.ENsetnodevalue(nodes, em.EN_BASEDEMAND, em.ENgetnodevalue(nodes, em.EN_BASEDEMAND)/valor |> Float64)
end

function muda_rugosidade(link::Int64, r::Float64)::Nothing
    em.ENsetlinkvalue(link, em.EN_ROUGHNESS, r)
end

# função objetivo
function objetivo(values::Array{Float64,1}, net)::Float64
    try
        erro::Float64 = 0.0
        update_network_values(values, net)
        for df in groupby(net.real_values, :vazao)
            muda_vazao.(df[1, :vazao], net.all_nodes)
            em.ENsolveH()
            for i in 1:size(df)[1]
                erro += (df[i,:pressure] - em.ENgetnodevalue(df[i,:node], em.EN_PRESSURE))^2
            end
            reverte_vazao.( df[1,:vazao], net.all_nodes)
        end
        return erro
    catch
        return Inf
    end
end


function get_dist(x, net)::Float64
    return sum((net.target-x).^2)^0.5
end

function get_groups(dim::Int64, tubulacoes)::Array{Array{Int64,1},1}
    groups = Array{Array{String,1},1}()
    if length(tubulacoes)%dim == 0
        count = 0
        for i in 1:dim
            append!(groups, [tubulacoes[count+1:Int(count+length(tubulacoes)/dim |> round)]])
            count += Int(length(tubulacoes)/dim |> round)
        end
    else
        count = 0
        for i in 1:dim
            if i==dim
                append!(groups, [tubulacoes[count+1:end]])
            else
                append!(groups, [tubulacoes[count+1:Int(count+length(tubulacoes)/dim |> round)]])
                count += Int(length(tubulacoes)/dim  |> round)
            end  
        end  
    end
    index_groups = Array{Array{Int64,1},1}()
    for group in groups
        append!(index_groups, [em.ENgetlinkindex.(group)])
    end
    return index_groups
end

function get_target_nodes(dim::Int64, tubulacoes, posicao::Float64)::Array{Int64,1}
    groups = Array{Array{String,1},1}()
    if length(tubulacoes)%dim == 0
        count = 0
        for i in 1:dim
            append!(groups, [tubulacoes[count+1:Int(count+length(tubulacoes)/dim |> round)]])
            count += Int(length(tubulacoes)/dim |> round)
        end
    else
        count = 0
        for i in 1:dim
            if i==dim
                append!(groups, [tubulacoes[count+1:end]])
            else
                append!(groups, [tubulacoes[count+1:Int(count+length(tubulacoes)/dim |> round)]])
                count += Int(length(tubulacoes)/dim |> round)
            end    
        end
    end
    index_groups = Array{Array{Int64,1},1}()
    for group in groups
        append!(index_groups, [em.ENgetlinkindex.(group)])
    end
    return [em.ENgetlinknodes(x[Int(length(x)*posicao |> round)])[1] for x in index_groups]
end

function get_real_values(paths, vazoes, target_rugo, 
    target_nodes, g_links, all_nodes)::Nothing
    if !isfile(paths.values)
        v = []
        n = []
        pressure = []
        update_network_values(target_rugo, g_links)
        for vazao in vazoes
            muda_vazao.(vazao, all_nodes)
            em.ENsolveH()
            for node in target_nodes
                append!(v, vazao)
                append!(n, node)
                append!(pressure, em.ENgetnodevalue(node, em.EN_PRESSURE))
            end
            reverte_vazao.(vazao, all_nodes)
        end
        df = DataFrame(node=n, vazao=v, pressure=pressure)
        CSV.write(paths.values, df)
    end
end


end # end epanet module

File: src_py/simulated_annealing/epanet.py
import pandas as pd
import epamodule as em
import numpy as np
import os 

class RealValues(object):
    def __init__(self, path_links, target_rugo, vazoes = [20,30,50,55,60,70]):
        self.dim = len(target_rugo)
        self.inp = path_links[2]
        self.links = open(path_links[1]).read().split('\n')
        self.saida = path_links[3]
        self.start_sim()
        self.nodes = self.get_nodes(path_links[0])
        self.get_groups()
        self.target_nodes = [em.ENgetlinknodes(x)[0] for x in self.get_target_links()]
        #self.target_nodes = [em.ENgetlinknodes(x)[0] for x in target_links]
        self.target_rugo = target_rugo
        self.vazoes = vazoes
    
    def start_sim(self):
        em.ENopen(self.inp)
        em.ENopenH()
    
    def close_sim(self):
        em.ENcloseH()
        em.ENclose()
    
    def restart(self):
        self.close_sim() 
        self.start_sim() 
    
    def muda_vazao(self, vazao):
        for node in self.nodes:
            em.ENsetnodevalue(node, em.EN_BASEDEMAND, em.ENgetnodevalue(node, em.EN_BASEDEMAND)*vazao)
    
    def reverte_vazao(self, vazao):
        for node in self.nodes:
            em.ENsetnodevalue(node, em.EN_BASEDEMAND, em.ENgetnodevalue(node, em.EN_BASEDEMAND)/vazao)

    def muda_rugosidade(self, no_link, rugosidade):
        for link in self.groups[no_link]:
            em.ENsetlinkvalue(link, em.EN_ROUGHNESS, rugosidade)
    
    def update_network_values(self, values):
        for i in range(len(values)):
            self.muda_rugosidade(i, values[i])
        em.ENsolveH()

    def get_nodes(self,path, test_value=1):
        arq = open(path)
        s = arq.read().split('\n')
        nodes = []
        for e in s:
            #try:
                #em.ENsetnodevalue(node, em.EN_BASEDEMAND, test_value)
            nodes.append(em.ENgetnodeindex(e))
            #except:
             #   print(f'{e} isnt a node')
        self.restart()
        return nodes 

    def get_groups(self):
        groups = []
        dim = self.dim
        if len(self.links)%dim ==0:
            count = 0
            for i in range(dim):
                groups.append(self.links[count:int(count+len(self.links)/dim)])
                count += int(len(self.links)/dim)
        else:
            count = 0
            for i in range(dim):
                if i==dim-1:
                    groups.append(self.links[count:])
                else:
                    groups.append(self.links[count:int(count+len(self.links)/dim)])
                    count += int(len(self.links)/dim)
                    
        #getid = np.vectorize(em.ENgetlinkindex) 
        #self.groups = getid(groups)
        index_groups = []
        for group in groups:
            index_groups.append([em.ENgetlinkindex(e) for e in group])
        self.groups = index_groups
        return index_groups

    def get_target_links(self):
        t_links = []
        for i in self.groups:
            t_links.append(i[0])
        return t_links


    def getRealValue(self):
        if not os.path.isfile(self.saida):
            values = []
            self.update_network_values(self.target_rugo)
            for vazao in self.vazoes:
                self.muda_vazao(vazao)
                for node in self.target_nodes:
                    em.ENsolveH()
                    pressure = em.ENgetnodevalue(node, em.EN_PRESSURE)
                    values.append([vazao, node, pressure])
                self.reverte_vazao(vazao)
            df = pd.DataFrame(np.array(values), columns = ['vazao', 'node', 'pressure'])
            df.loc[:,'node'] = df['node'].astype(int)
            df.to_csv(self.saida, index=False)



class Rede(object):
    def __init__(self, l_links, group_links, t, bound):
        self.bounds = bound
        self.limita = np.vectorize(lambda x: bound[0] if x< bound[0] else bound[1] if x>bound[1] else x)
        self.nodes = l_links[0]
        self.links = l_links[1]
        self.inp = l_links[2]
        print("Começando simulação")
        self.start_sim()
        self.target = t
        self.valores_reais = pd.read_csv(l_links[3]) # dicionario(vazao -> [valores de pressão pra cada grupo])
        self.nodes = self.get_nodes()
        self.g_links = group_links # vetor[nº grupo] = [link1, link2, .....] começando do 0
    
    def start_sim(self):
        em.ENopen(self.inp)
        em.ENopenH()
    
    def close_sim(self):
        em.ENcloseH()
        em.ENclose()
    
    def restart(self):
        self.close_sim() 
        self.start_sim()   


    def get_nodes(self):
        arq = open(self.nodes)
        s = arq.read().split('\n')
        nodes = [em.ENgetnodeindex(e) for e in s]
        return nodes 
    


    def muda_vazao(self, vazao):
        for node in self.nodes:
            em.ENsetnodevalue(node, em.EN_BASEDEMAND, em.ENgetnodevalue(node, em.EN_BASEDEMAND)*vazao)
    
    def reverte_vazao(self, vazao):
        for node in self.nodes:
            em.ENsetnodevalue(node, em.EN_BASEDEMAND, em.ENgetnodevalue(node, em.EN_BASEDEMAND)/vazao)

    def muda_rugosidade(self, no_link, rugosidade):
        for link in self.g_links[no_link]:
            em.ENsetlinkvalue(link, em.EN_ROUGHNESS, rugosidade)
    
    def update_network_values(self, values):
        for i in range(len(values)):
            self.muda_rugosidade(i, values[i])
        em.ENsolveH()
    
    def objetivo(self, values):
        #values = self.limita(values)
        #values = np.array(values)
        #if sum(values > self.bounds[1])+sum(values < self.bounds[0]) > 0:
        #    return 1000
        #print(values)
        try:
            self.update_network_values(values)
            erro = 0
            for vazao in self.valores_reais['vazao'].unique():
                self.muda_vazao(vazao)
                em.ENsolveH()
                for v in self.valores_reais[self.valores_reais['vazao']==vazao].values:
                    erro += (v[2]-em.ENgetnodevalue(int(v[1]), em.EN_PRESSURE))**2
                self.reverte_vazao(vazao)
            self.ultimo_ponto = values
        #self.restart()
        finally:
            return erro
        return erro

    def gradient(self, x, h=0.0001): #vetor ponto
        g = np.zeros((len(x)))
        dim = len(x)
        for i in range(dim):
            #derivada_parcial = (self.objetivo(x) - self.objetivo([x[e]-h if e==i else x[e] for e in range(dim)]))/h
            #derivada_parcial += (self.objetivo([x[e]+h if e==i else x[e] for e in range(dim)]) - self.objetivo(x))/h
            derivada_parcial = (self.objetivo([x[e]+h if e==i else x[e] for e in range(dim)])-self.objetivo([x[e]-h if e==i else x[e] for e in range(dim)]))/(2*h)
            #g[i] += round(derivada_parcial,3)
            g[i] += round(derivada_parcial,3)
            
        return g

    def get_dist(self, p):
        return np.linalg.norm(self.target-p)


class RealValuesNos(object):
    def __init__(self, path_links, target_rugo, nos_dim = 30, posicao = 0, vazoes = [20,30,50,55,60,70]):
        self.nos_dim = nos_dim
        self.posicao = posicao # em %
        self.dim = len(target_rugo)
        self.inp = path_links[2]
        self.links = open(path_links[1]).read().split('\n')
        self.saida = path_links[3]
        self.start_sim()
        self.nodes = self.get_nodes(path_links[0])
        self.get_groups()
        self.target_nodes = self.get_target_nodes()
        #self.target_nodes = [em.ENgetlinknodes(x)[0] for x in target_links]
        self.target_rugo = target_rugo
        self.vazoes = vazoes
        
        
        
    
    def start_sim(self):
        em.ENopen(self.inp)
        em.ENopenH()
    
    def close_sim(self):
        em.ENcloseH()
        em.ENclose()
    
    def restart(self):
        self.close_sim() 
        self.start_sim() 
    
    def muda_vazao(self, vazao):
        for node in self.nodes:
            em.ENsetnodevalue(node, em.EN_BASEDEMAND, em.ENgetnodevalue(node, em.EN_BASEDEMAND)*vazao)
    
    def reverte_vazao(self, vazao):
        for node in self.nodes:
            em.ENsetnodevalue(node, em.EN_BASEDEMAND, em.ENgetnodevalue(node, em.EN_BASEDEMAND)/vazao)

    def muda_rugosidade(self, no_link, rugosidade):
        for link in self.groups[no_link]:
            em.ENsetlinkvalue(link, em.EN_ROUGHNESS, rugosidade)
    
    def update_network_values(self, values):
        for i in range(len(values)):
            self.muda_rugosidade(i, values[i])
        em.ENsolveH()

    def get_nodes(self,path, test_value=1): # pega todos os nós
        arq = open(path)
        s = arq.read().split('\n')
        nodes = []
        for e in s:
            #try:
                #em.ENsetnodevalue(node, em.EN_BASEDEMAND, test_value)
            nodes.append(em.ENgetnodeindex(e))
            #except:
             #   print(f'{e} isnt a node')
        self.restart()
        return nodes 

    def get_groups(self): # pega os grupos de tubulações de acordo com a dimensão
        groups = []
        dim = self.dim
        if len(self.links)%dim ==0:
            count = 0
            for i in range(dim):
                groups.append(self.links[count:int(count+len(self.links)/dim)])
                count += int(len(self.links)/dim)
        else:
            count = 0
            for i in range(dim):
                if i==dim-1:
                    groups.append(self.links[count:])
                else:
                    groups.append(self.links[count:int(count+len(self.links)/dim)])
                    count += int(len(self.links)/dim)
        #getid = np.vectorize(em.ENgetlinkindex) 
        #self.groups = getid(groups)
        index_groups = []
        for group in groups:
            index_groups.append([em.ENgetlinkindex(e) for e in group])
        self.groups = index_groups
        return index_groups

    def get_target_nodes(self): # pega 30 nós de acordo com a divisão de nos_dim grupos de  tubulações
        groups = []
        dim = self.nos_dim
        if len(self.links)%dim ==0:
            count = 0
            for i in range(dim):
                groups.append(self.links[count:int(count+len(self.links)/dim)])
                count += int(len(self.links)/dim)
        else:
            count = 0
            for i in range(dim):
                if i==dim-1:
                    groups.append(self.links[count:])
                else:
                    groups.append(self.links[count:int(count+len(self.links)/dim)])
                    count += int(len(self.links)/dim)
        index_groups = []
        for group in groups:
            index_groups.append([em.ENgetlinkindex(e) for e in group])
        return [em.ENgetlinknodes(x[int(len(x)*self.posicao)])[0] for x in index_groups]

    def get_target_links(self):
        t_links = []
        for i in self.groups:
            t_links.append(i[0])
        return t_links


    def getRealValue(self):
        if not os.path.isfile(self.saida):
            values = []
            self.update_network_values(self.target_rugo)
            for vazao in self.vazoes:
                self.muda_vazao(vazao)
                for node in self.target_nodes:
                    em.ENsolveH()
                    pressure = em.ENgetnodevalue(node, em.EN_PRESSURE)
                    values.append([vazao, node, pressure])
                self.reverte_vazao(vazao)
            df = pd.DataFrame(np.array(values), columns = ['vazao', 'node', 'pressure'])
            df.loc[:,'node'] = df['node'].astype(int)
            df.to_csv(self.saida, index=False)
